### PR TITLE
Re-order and refactor help for client/server validation

### PR DESCRIPTION
**Pull request by pljones**
**Short description of changes**

Review and refactor command line handling in `src/main.cpp` to standardise how it all works.

**Context: Fixes an issue?**

Was part of #1867 (persistent directory lists) but really is nothing to do with it.  Might unnecessarily cause conflicts in that change should `src/main.cpp` be amended by some other change, if left as part of it.

There's what appears to be a lot of code change - in fact, it's mostly moving existing code around.  The main _changes_ are from (new) line 533 onwards.

**Does this change need documentation? What needs to be documented and how?**

The code change should be self-documenting, I hope.

This change affects the `--help` output of Jamulus, so any documetation that references that will need reviewing.  I have not looked into that aspect.

**Status of this Pull Request**

Working, tested code.

**What is missing until this pull request can be merged?**

Needs reviewing as an independent pull request.  Had been reviewed under #1867 by @ann0see, I believe.  See https://github.com/jamulussoftware/jamulus/pull/1867#pullrequestreview-687932186

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

### CODE REVIEW

**Comment by ann0see** [comment]
Put on my todo list but I can’t guarantee anything

**Comment by ann0see** [comment]
> This change affects the --help output of Jamulus, so any documetation that references that will need reviewing. I have not looked into that aspect.

I think it’s not really used anywhere yet. The CLI page was and is created manually.

**Review by ann0see — COMMENTED** [review]
Ran "./Jamulus -e" which produced the error message:
`./Jamulus: '--directoryserver' needs a string argument.`
This might be confusing since we passed -e not --directoryserver. Also the error message with server only options might need some thoughts and tests (In some cases I can't remember "Did you omit --server" was confusing. Maybe `./Jamulus -e -c`?). 

All in all it seems to be quite OK?

File: src/main.cpp [hunk 1]
```diff
@@ -105,144 +107,131 @@ int main ( int argc, char** argv )
...
-                               i,
-                               "--showallservers", // no short form
-                               "--showallservers" ) )
+        // Central server ** D E P R E C A T E D ** ----------------------------
+        if ( GetStringArgument ( argc,
+                                 argv,
+                                 i,
+                                 "--centralserver", // no short form
+                                 "--centralserver",
+                                 strArgument ) )
```
[the remark this answers is not in the document]
Comment (pljones) [reply src/main.cpp:209]: The `-e` giving `--directoryserver` in the error is an existing feature (has been for a long time, I think).

I also don't see how this is "inconsistent".  In what way?  It's not changed?

File: src/main.cpp [hunk 2]
```diff
@@ -280,20 +262,42 @@ int main ( int argc, char** argv )
...
+        // Server Public IP ----------------------------------------------------
         if ( GetStringArgument ( argc,
                                  argv,
                                  i,
-                                 "--clientname", // no short form
-                                 "--clientname",
+                                 "--serverpublicip", // no short form
+                                 "--serverpublicip",
                                  strArgument ) )
         {
```
[the remark this answers is not in the document]
Comment (pljones) [reply src/main.cpp:286]: Again, not changed.

File: src/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ann0see) [reply src/main.cpp:209]: It’s at least confusing since the user used -e not directoryserver. But if it’s not new, that’s unrelated.

**Review by ann0see — APPROVED** [review]

**Review by softins — APPROVED** [review]
Sorry for the delay - Jamulus is taking a back seat at the moment.

This PR looks ok to me. I agree with @ann0see's comment https://github.com/jamulussoftware/jamulus/pull/1896#pullrequestreview-697517772 regarding the error message. This could be addressed with the following diff:
```
diff --git a/src/main.cpp b/src/main.cpp
index 0a0a6fcb..e58d5e17 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -942,7 +942,7 @@ bool GetStringArgument ( int argc, char** argv, int& i, QString strShortOpt, QSt
     {
         if ( ++i >= argc )
         {
-            qCritical() << qUtf8Printable ( QString ( "%1: '%2' needs a string argument." ).arg ( argv[0] ).arg ( strLongOpt ) );
+            qCritical() << qUtf8Printable ( QString ( "%1: '%2' needs a string argument." ).arg ( argv[0] ).arg ( argv[i-1] ) );
             exit ( 1 );
         }

@@ -970,7 +970,7 @@ bool GetNumericArgument ( int     argc,
         QString errmsg = "%1: '%2' needs a numeric argument between '%3' and '%4'.";
         if ( ++i >= argc )
         {
-            qCritical() << qUtf8Printable ( errmsg.arg ( argv[0] ).arg ( strLongOpt ).arg ( rRangeStart ).arg ( rRangeStop ) );
+            qCritical() << qUtf8Printable ( errmsg.arg ( argv[0] ).arg ( argv[i-1] ).arg ( rRangeStart ).arg ( rRangeStop ) );
             exit ( 1 );
         }

@@ -978,7 +978,7 @@ bool GetNumericArgument ( int     argc,
         rValue = strtod ( argv[i], &p );
         if ( *p || ( rValue < rRangeStart ) || ( rValue > rRangeStop ) )
         {
-            qCritical() << qUtf8Printable ( errmsg.arg ( argv[0] ).arg ( strLongOpt ).arg ( rRangeStart ).arg ( rRangeStop ) );
+            qCritical() << qUtf8Printable ( errmsg.arg ( argv[0] ).arg ( argv[i-1] ).arg ( rRangeStart ).arg ( rRangeStop ) );
             exit ( 1 );
         }
```

Then the error message would mention the actual option supplied. But I don't feel strongly enough to block the PR for it.

**Comment by pljones** [comment]
Patched:
```
peter@fs-peter:~/git/Jamulus-wip$ ./Jamulus -e
./Jamulus: '-e' needs a string argument.
```

**Pull request merged**